File: rec-core/src/main/java/net/kimleo/rec/v2/accessor/Accessor.java
package net.kimleo.rec.v2.accessor;

import net.kimleo.rec.Pair;
import net.kimleo.rec.v2.accessor.lexer.Lexer;
import net.kimleo.rec.concept.Accessible;

import java.util.Map;

public class Accessor<T> {

    private final Map<String, Integer> fieldMap;
    private final Integer leastCapacity;

    public Accessor(String[] fields) {
        Pair<Map<String, Integer>, Integer> fieldMapPair = Lexer.buildFieldMapPair(fields);
        fieldMap = fieldMapPair.getFirst();
        leastCapacity = fieldMapPair.getSecond();
    }

    public Map<String, Integer> getFieldMap() {
        return fieldMap;
    }

    public Integer getLeastCapacity() {
        return leastCapacity;
    }

    public RecordWrapper<T> create(Accessible<T> record) {
        assert (record.size() >= leastCapacity);

        return new RecordWrapper<>(fieldMap, record);
    }

    public RecordWrapper<T> of(Accessible<T> record) {
        return create(record);
    }
}

File: src/main/kotlin/net/kimleo/rec/accessor/Accessor.kt
package net.kimleo.rec.accessor

import net.kimleo.rec.accessor.lexer.Lexer
import net.kimleo.rec.record.Record
import net.kimleo.rec.record.Field

class Accessor(fields: Array<out String>) {
    val fieldMapPair = Lexer.buildFieldMapPair(*fields)
    val fieldMap = fieldMapPair.first
    val leastCapacity = fieldMapPair.second

    constructor(record: Record): this(record.fields.map(Field::value).toTypedArray()) {}

    fun create(record: Record): RecordWrapper {
        assert(record.size >= leastCapacity)

        return RecordWrapper(fieldMap, record)
    }

    fun of(record: Record) = create(record)
}